lib/rules/require-valid-css-selector-in-test-helpers.js

```diff
@@ -0,0 +1,146 @@
+'use strict';
+
+const { getSourceModuleNameForIdentifier } = require('../utils/import');
+const { getPropertyValue, parseCallee, getAncestor } = require('../utils/utils');
+const { isImportDeclaration } = require('../utils/types');
+const cssTree = require('css-tree');
+
+const TEST_HELPER_IMPORTS = new Set(['fillIn', 'findAll', 'find', 'click', 'focus']);
+const QUERY_SELECTOR_METHODS = new Set(['querySelectorAll', 'querySelector']);
+const PARENT_NODE_NAMES = new Set(['element', 'document']);
+const SELECTOR_RULES = Object.freeze({
+  unclosedAttr: {
+    hasError: (selector) => selector.includes('[') && !selector.includes(']'),
+    fix(node, selector, fixer) {
+      return fixer.replaceText(node.arguments[0], `'${selector}]'`);
+    },
+    errorMessage:
+      'Syntax error, you used an unclosed attribute selector: "{{selector}}", should be: "{{selector}}]"',
+  },
+  idStartsWithNumber: {
+    hasError: (selector) => selector.match(/^#\d/),
+    fix() {},
+    errorMessage: 'Syntax error, ids cannot start with a number: "{{selector}}"',
+  },
+  other: {
+    hasError: (selector) => !_isValidSelector(selector),
+    fix() {},
+    errorMessage: 'Syntax error, "{{selector}}" is not a valid selector',
+  },
+});
+
+function _isValidSelector(selector) {
+  try {
+    cssTree.parse(selector, {
+      context: 'selector',
+    });
+  } catch {
+    return false;
+  }
+
+  return true;
+}
+
+function _isAssertDomCall(node, context) {
+  const calleeName = parseCallee(node) || [];
+
+  if (calleeName[1] !== 'dom') {
+    return false;
+  }
+
+  const testCallExpressionNode = getAncestor(
+    node,
+    (ancestorNode) =>
+      getPropertyValue(ancestorNode, 'callee.name') === 'test' &&
+      getSourceModuleNameForIdentifier(context, ancestorNode) === 'qunit'
+  );
+
+  if (!testCallExpressionNode) {
+    return false;
+  }
+
+  return calleeName[0] === getPropertyValue(testCallExpressionNode, 'arguments.1.params.0.name');
+}
+
+function _isQuerySelectorCall(node, context) {
+  const calleeName = parseCallee(node);
+
+  const testCallExpressionNode = getAncestor(
+    node,
+    (ancestorNode) =>
+      getPropertyValue(ancestorNode, 'callee.name') === 'test' &&
+      getSourceModuleNameForIdentifier(context, ancestorNode) === 'qunit'
+  );
+
+  if (!testCallExpressionNode) {
+    return false;
+  }
+
+  return QUERY_SELECTOR_METHODS.has(calleeName[1]) && PARENT_NODE_NAMES.has(calleeName[0]);
+}
+
+function _isTestHelperCall(node, context) {
+  const calleeName = parseCallee(node) || [];
+
+  const [program] = context.getAncestors(node);
```

Comment:
It's pretty inefficient for us to have to look through all the ancestors for import declarations every time we see a function call. Instead, we should use the technique that most other rules follow in their `ImportDeclaration(node)` functions, which is to gather the imported function names upfront, and then perform a simple check to see if the current function call matches the imported function name.

This will also make it easier to handle imported functions that are renamed (`import { foo as fooRenamed } ...`).

Reply:
Done